src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,599 @@
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    /**
+     * Instantiates a KDF object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param algParameterSpec
+     *     the algorithm parameters
+     */
+    private KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                AlgorithmParameterSpec algParameterSpec) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        // the lock is not needed, because the Spi will already be set in
+        // chooseProvider
+        lock = null;
+    }
+
+    private KDF(Service s, Iterator<Service> t, String algorithm,
+                AlgorithmParameterSpec algParameterSpec) {
+        firstService = s;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        lock = new Object();
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        chooseFirstProvider();
+        return provider.getName();
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDF} implementation for the
+     *     specified algorithm
+     * @throws NullPointerException
+     *     if {@code algorithm} is {@code null}
+     */
+    public static KDF getInstance(String algorithm)
+        throws NoSuchAlgorithmException {
+        try {
+            return getInstance(algorithm, (AlgorithmParameterSpec) null);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "Received an InvalidAlgorithmParameterException. Does this "
+                + "algorithm require an AlgorithmParameterSpec?", e);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm from
+     * the specified security provider.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     * @param provider
+     *     the provider to use for this key derivation; if null, this method is
+     *     equivalent to {@code getInstance(String)}
```

Comment:
It might be better to disallow null here. If the code doesn't have a provider name then it would be clearer to use the 1-arg getInstance method. That would also help catch bugs where the provider name is null due to some bug.

Reply:
I agree that is probably better. Our JCE APIs are somewhat inconsistent on this.